File: utils/calculations.py
"""Transparent rules used to simulate operational intelligence indicators."""

from __future__ import annotations

import pandas as pd


def health_category(score: float) -> str:
    """Classify equipment health using the documented prototype thresholds."""
    if score >= 80:
        return "Healthy"
    if score >= 60:
        return "Warning"
    return "Critical"


def equipment_health_counts(equipment: pd.DataFrame) -> dict[str, int]:
    """Return the number of assets in each health band."""
    categories = equipment["health_score"].apply(health_category)
    counts = categories.value_counts()
    return {name: int(counts.get(name, 0)) for name in ("Healthy", "Warning", "Critical")}


def dashboard_kpis(
    vessels: pd.DataFrame,
    voyages: pd.DataFrame,
    equipment: pd.DataFrame,
    alerts: pd.DataFrame,
    safety: pd.DataFrame,
) -> dict[str, str | int]:
    """Calculate executive KPIs from simple, explainable prototype rules."""
    active_vessels = int(vessels["operational_status"].isin(["Underway", "At berth", "In Port", "At Anchorage"]).sum())
    critical_alerts = int(((alerts["severity"] == "Critical") & (alerts["status"] != "Closed")).sum())
    predicted_failures = int((equipment["predicted_failure_days"].between(0, 30)).sum())
    delayed_voyages = int((voyages["predicted_eta"] > voyages["planned_eta"]).sum())
    potential_saving = float(voyages["potential_fuel_saving_pct"].mean())
    high_risk_safety = int(((safety["risk_level"] == "High") & (safety["status"] != "Closed")).sum())
    return {
        "Active vessels": active_vessels,
        "Critical alerts": critical_alerts,
        "Predicted failures (30d)": predicted_failures,
        "Delayed voyages": delayed_voyages,
        "Potential fuel saving": f"{potential_saving:.1f}%",
        "Open high-risk safety": high_risk_safety,
    }


def maintenance_summary(equipment: pd.DataFrame) -> dict[str, int]:
    """Summarise scheduled and urgent maintenance work."""
    return {
        "Due in 7 days": int((equipment["maintenance_due_days"] <= 7).sum()),
        "Overdue": int((equipment["maintenance_due_days"] < 0).sum()),
        "Work orders open": int((equipment["maintenance_status"] != "Complete").sum()),
    }

File: src/utils/calculations.ts
import { Alert, Equipment, SafetyEvent, Vessel, Voyage } from '../types/maritime';

export function healthCategory(score: number): 'Healthy' | 'Warning' | 'Critical' {
  if (score >= 80) return 'Healthy';
  if (score >= 60) return 'Warning';
  return 'Critical';
}

export function equipmentHealthCounts(equipment: Equipment[]): {
  Healthy: number;
  Warning: number;
  Critical: number;
} {
  const counts = { Healthy: 0, Warning: 0, Critical: 0 };
  for (const item of equipment) {
    const cat = healthCategory(item.health_score);
    counts[cat]++;
  }
  return counts;
}

export function dashboardKpis(
  vessels: Vessel[],
  voyages: Voyage[],
  equipment: Equipment[],
  alerts: Alert[],
  safety: SafetyEvent[]
): Record<string, string | number> {
  const activeVessels = vessels.filter((v) =>
    ['Underway', 'At berth', 'In Port', 'At Anchorage'].includes(v.operational_status)
  ).length;

  const criticalAlerts = alerts.filter(
    (a) => a.severity === 'Critical' && a.status !== 'Closed'
  ).length;

  const predictedFailures = equipment.filter(
    (e) => e.predicted_failure_days >= 0 && e.predicted_failure_days <= 30
  ).length;

  const delayedVoyages = voyages.filter(
    (voy) => new Date(voy.predicted_eta) > new Date(voy.planned_eta)
  ).length;

  const potentialSaving =
    voyages.length > 0
      ? voyages.reduce((acc, voy) => acc + (voy.potential_fuel_saving_pct || 0), 0) / voyages.length
      : 0;

  const highRiskSafety = safety.filter(
    (s) => s.risk_level === 'High' && s.status !== 'Closed'
  ).length;

  return {
    'Active vessels': activeVessels,
    'Critical alerts': criticalAlerts,
    'Predicted failures (30d)': predictedFailures,
    'Delayed voyages': delayedVoyages,
    'Potential fuel saving': `${potentialSaving.toFixed(1)}%`,
    'Open high-risk safety': highRiskSafety,
  };
}

export function maintenanceSummary(equipment: Equipment[]): {
  'Due in 7 days': number;
  Overdue: number;
  'Work orders open': number;
} {
  return {
    'Due in 7 days': equipment.filter((e) => e.maintenance_due_days <= 7).length,
    Overdue: equipment.filter((e) => e.maintenance_due_days < 0).length,
    'Work orders open': equipment.filter((e) => e.maintenance_status !== 'Complete').length,
  };
}
